### PR TITLE
Add an option to use the input model bytes (ORT format only) directly without copy at session creation

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: Add an option to use the input model bytes (ORT format only) directly without copy at session creation

**Motivation and Context**
- When we load an ORT format model from given byte array, we copy the array first to ensure the bytes are valid. There is a recent issue #8328, where the current behavior may cause extra memory usage where iOS may kill the process
- Add an option to not copy the input model bytes, however, in this case, the user will have to guarantee the buffer is valid until the session is created (this is the 1st step to reduce memory consumption). The default behavior is still copying the model bytes. The user has to specifically set the option to disable the copy.
- Added UT cases

### CODE REVIEW

File: onnxruntime/core/session/inference_session.h [hunk 1]
```diff
@@ -704,13 +704,29 @@ class InferenceSession {
...
+  //   and some for the Initialize (create SessionState).
   // Short term we free them after Initialize.
   // Longer term we may want to directly refer to offsets in this buffer for initializers so we don't need to copy
   // those into new OrtValue instances, at which point we won't free them until the InferenceSession goes away.
-  std::vector<uint8_t> ort_format_model_bytes_;
+  gsl::span<const uint8_t> ort_format_model_bytes_;
+
+  // This holds the actual model data when the session is started with a model_uri
+  // In case if the session is started with an input byte array contains model data, this will be empty
+  std::vector<uint8_t> ort_format_model_bytes_data_holder_;
```
Comment (skottmckay) [inline onnxruntime/core/session/inference_session.h:729]: nit: may be empty?

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h [hunk 2]
```diff
@@ -61,6 +61,13 @@ static const char* const kOrtSessionOptionsUseDeviceAllocatorForInitializers = "
 static const char* const kOrtSessionOptionsConfigAllowInterOpSpinning = "session.inter_op.allow_spinning";
 static const char* const kOrtSessionOptionsConfigAllowIntraOpSpinning = "session.intra_op.allow_spinning";
 
+// Key for disable copy model bytes for ORT format
+// By default we will copy the model bytes for ORT format at the time of session creation to ensure the model bytes
+// buffer is valid.
+// Setting this option to "1" will disable copy the model bytes. The caller has to guarantee that the model bytes
+// are valid until the ORT session using the model bytes is destroyed.
+static const char* const kOrtSessionOptionsConfigDisableCopyORTModelBytes = "session.disable_copy_ort_model_bytes";
+
```
Comment (skottmckay) [inline include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:70]: nit: would 'use_ort_model_bytes_directly' be slightly more generic and cover both the copy and potential usage of initializers from the bytes? not a big difference though.

Comment (skottmckay) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:70]: Can you also update [https://www.onnxruntime.ai/docs/how-to/mobile/model-execution.html](https://www.onnxruntime.ai/docs/how-to/mobile/model-execution.html) to call out this option?

**Review by skottmckay — DISMISSED** [review]
:shipit:

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h (hunk 2 above)
Comment (gwang-msft) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:71]: > Can you also update https://www.onnxruntime.ai/docs/how-to/mobile/model-execution.html to call out this option?

Will do in a separated PR

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h [hunk 3]
```diff
@@ -71,5 +79,3 @@ static const char* const kOrtSessionOptionsConfigAllowIntraOpSpinning = "session
 // If not specified, the default set of stop ops is used. To specify an empty stop ops types list and disable stop op
 // exclusion, set the value to "".
 static const char* const kOrtSessionOptionsConfigNnapiEpPartitioningStopOps = "ep.nnapi.partitioning_stop_ops";
-
-// NNAPI EP keys end
```
Comment (edgchen1) [inline include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:75]: if removing the "end" comment, can you also remove the "begin" one and change the note to say "The NNAPI EP options..." instead?

**Review by edgchen1 — APPROVED** [review]

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h (hunk 3 above)
Comment (gwang-msft) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:75]: Will do in my next change

**Pull request merged**